feat: copy directories recursively

fdschmidt93 opened the pull request:
Extends `Path:copy` to recursively copy folders for https://github.com/nvim-telescope/telescope.nvim/pull/1257

This is pretty much the plenary equivalent of how nvim_tree performs copying [here](https://github.com/kyazdani42/nvim-tree.lua/blob/d7f73b5ae9c8fa85535c32e2861c2cb97df5d56b/lua/nvim-tree/fs.lua#L146).

E: tests are TODO

Conni2461 commented on lua/plenary/path.lua:
```diff
@@ -463,10 +464,33 @@ function Path:copy(opts)
+    local scan = require "plenary.scandir"
+    scan.scan_dir(self.filename, {
+      respect_gitignore = F.if_nil(opts.respect_gitignore, true, opts.respect_gitignore),
+      hidden = F.if_nil(opts.hidden, false, opts.hidden),
```
i expect hidden files also to be copied.

Conni2461 commented on lua/plenary/path.lua:
```diff
@@ -463,10 +464,33 @@ function Path:copy(opts)
+        table.insert(data, Path:new(entry))
+      end,
+    })
+    for _, path_ in ipairs(data) do
```
i dont like this name `path_`

Conni2461 commented on lua/plenary/path.lua:
```diff
@@ -463,10 +464,33 @@ function Path:copy(opts)
+      depth = 1,
+      add_dirs = F.if_nil(opts.add_dirs, true, opts.add_dirs),
+      on_insert = function(entry)
+        table.insert(data, Path:new(entry))
```
why insert `Path:new` in a table. Why not use the return of scan_dir and then create new Path items in the for loop below?

Conni2461 reviewed (commented):
Some tests would be nice

https://github.com/nvim-lua/plenary.nvim/blob/master/tests/plenary/path_spec.lua#L354-L396

fdschmidt93 commented:
Ok, I've finally come around to improve this over my arguably bad initial implementation.

We now:
* Not overriding by default was an unexpected default, we to `override=true` like `cp`. We can revert that though
* Support `opts.interactive` with `vim.ui.select` which works in naive test, though unclear what happens if table is accessed right after `copy`
* `Path:copy` returns a table of `{[Path of dest] = bool}` where bool indicates whether the `destination` was copied to. This leads to nicely having nested tables when copying a directory

The second test is quite involved but we overall make sure that we:
* can(not) copy directories recursively
* whether hidden files are copied
* don't override if override is not set

fdschmidt93 commented:
@Conni2461 sorry, friendly ping :) could you please make time to review this PR prior to when the file browser will be deprecated?

Conni2461 commented:
Its on top of my list. I looked at it already and it looks mostly good but i need to verify a couple of things :grimacing: 
I've it done till friday morning. If i recall correctly we'll add deprecation messages for file_browser on friday, right?

fdschmidt93 commented:
Ok :) Yeah, Friday works for me. I'll prepare a PR then.

Conni2461 commented on lua/plenary/path.lua:
```diff
@@ -474,20 +474,77 @@ function Path:rename(opts)
+---@field destination string|Path: target file path to copy to
+---@field recursive bool: whether to copy folders recursively (default: true)
+---@field override bool: whether to override files (default: true)
+---@field interactive bool: confirm if copy would override; precedes `override` (default: false)
```
does it make sense to have interactive default true ? Does file_browser disable it ? Because if `vim.select` is set to telescope it will close it

Conni2461 commented on lua/plenary/path.lua:
```diff
@@ -474,20 +474,77 @@ function Path:rename(opts)
+---@field recursive bool: whether to copy folders recursively (default: true)
+---@field override bool: whether to override files (default: true)
+---@field interactive bool: confirm if copy would override; precedes `override` (default: false)
+---@field respect_gitignore bool: skip folders ignored by all detected `gitignore`s (default: true)
```
does it make sense to have respect_gitignore to true? `cp` doesnt have it either.

Conni2461 commented on lua/plenary/path.lua:
```diff
@@ -474,20 +474,77 @@ function Path:rename(opts)
+---@return table {[Path of destination]: bool} indicating success of copy; nested tables constitute sub dirs
 function Path:copy(opts)
   opts = opts or {}
+  opts.recursive = F.if_nil(opts.recursive, false, opts.recursive)
```
docs say recursive is true on default. implementation says false.

fdschmidt93 replied:
I've opted to `recursive=false` by default such that it's more in line with `cp`.

Conni2461 replied:
Yeah following `cp` is a good idea. I just "pointed" out minor inconsistencies in documentation and implementation.

Thanks for cleaning it up

Conni2461 commented on lua/plenary/path.lua:
```diff
@@ -474,20 +474,77 @@ function Path:rename(opts)
+        end
+      )
+    else
+      flags.excl = not opts.override
+      -- nil: not overriden if `override = false`
+      success[dest] = uv.fs_copyfile(self:absolute(), dest:absolute(), flags) or false
```
do we even need the flags table or can we just do it in the same `fs_copyfile` line ?

Conni2461 commented on lua/plenary/path.lua:
```diff
@@ -474,20 +474,77 @@ function Path:rename(opts)
+  -- dir
+  if opts.recursive then
+    dest:mkdir {
+      parents = F.if_nil(opts.parents, true, opts.parents),
+      exists_ok = F.if_nil(opts.exists_ok, true, opts.exists_ok),
```
these two opts arent documented. not sure if we even need to have them configurable

fdschmidt93 replied:
I've made them configurable as they stem from `mkdir` and I feel `copy` should then just follow.

fdschmidt93 replied:
For e2: I've modified the `Path:mkdir` implementation to skip `parents` etc if the folder already exists which fixed the tests right away.

Conni2461 replied:
good catch thanks

Conni2461 reviewed (approved):
LGTM thanks :)

Just some minor questions :) can you also rebase ?

fdschmidt93 commented:
Thanks for the review :) Finally came around to addressing the comments. I've also streamlined a bit to more closely follow `cp` because I presume that's what the majority of users would expect.

Conni2461 reviewed (approved):
Thanks :)
